### PR TITLE
[3/4] Intel GPU Runtime Upstreaming for Device

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118613
* #118528
* #118523
* #118091
* #117734
* #117619
* #117611
* #116869
* #118846
* __->__ #116850

# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), As mentioned in [[RFC] Intel GPU Runtime Upstreaming](https://github.com/pytorch/pytorch/issues/114842), this third PR  covers the changes under `libtorch_python`.

# Design
This PR primarily offers device-related APIs in python frontend, including
- `torch.xpu.is_available`
- `torch.xpu.device_count`
- `torch.xpu.current_device`
- `torch.xpu.set_device`
- `torch.xpu.device`
- `torch.xpu.device_of`
- `torch.xpu.get_device_name`
- `torch.xpu.get_device_capability`
- `torch.xpu.get_device_properties`
- ====================
- `torch.xpu._DeviceGuard`
- `torch.xpu._is_compiled`
- `torch.xpu._get_device`

# Additional Context
We will implement the support of lazy initialization in the next PR. 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: torch/random.py [hunk 1]
```diff
@@ -43,9 +43,6 @@ def manual_seed(seed) -> torch._C.Generator:
     if not torch.mps._is_in_bad_fork():
         torch.mps.manual_seed(seed)
 
-    if hasattr(torch, 'xpu') and not torch.xpu._is_in_bad_fork():
-        torch.xpu.manual_seed_all(seed)
-
```
Comment (EikanWang) [inline torch/random.py:48]: Why does XPU not need to set a manual seed?

File: torch/random.py [hunk 2]
```diff
@@ -65,9 +62,6 @@ def seed() -> int:
...
```
Comment (EikanWang) [inline torch/random.py:70]: ditto

File: torch/csrc/api/include/torch/all.h [hunk 3]
```diff
@@ -21,3 +21,4 @@
 #include <torch/types.h>
 #include <torch/utils.h>
 #include <torch/version.h>
+#include <torch/xpu.h>
```
Comment (EikanWang) [inline torch/csrc/api/include/torch/all.h:24]: If users do not build XPU(USE_XPU=0), what's the behavior?

**Review by EikanWang — CHANGES_REQUESTED** [review]

File: torch/random.py (hunk 1 above)
Comment (guangyey) [reply torch/random.py:48]: Here will raise an error since `hasattr(torch, 'xpu')` always `True`. But currently `manual_seed_all` and `_is_in_bad_fork` don't exist in torch.xpu module. This error will be triggered via running UT `test_xpu.py`.
We will add them back in PR about `Generator`.

File: torch/csrc/api/include/torch/all.h (hunk 3 above)
Comment (guangyey) [reply torch/csrc/api/include/torch/all.h:24]: It is Ok if USE_XPU=0 since the function in xpu.h is implemented via hooks.

File: torch/random.py (hunk 1 above)
Comment (EikanWang) [reply torch/random.py:48]: I see. So I wonder if you can add some informative comments here to help reviewers to understand the background.

File: torch/random.py (hunk 2 above)
Comment (guangyey) [reply torch/random.py:70]: We remove this code here, and add them back with PR about `Generator`.

**Review by EikanWang — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

**Review by gujinghui — APPROVED** [review]

**Comment by EikanWang** [comment]
Regarding the failed case `test_autograd.py::TestAutograd::test_checkpointing_without_reentrant_dataparallel`, it is not applicable for XPU now. Because this test case counts on XPU aten operations, while the XPU aten operations are WIP. We will submit a PR to skip this test case just like mps, rocm.

**Comment by EikanWang** [comment]
@malfet , https://github.com/pytorch/pytorch/pull/117621 has landed to bypass the test cases that are not applicable for XPU at this moment and XPU CI can pass now. May I know if you have any comments on this PR?

File: torch/xpu/__init__.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
+from functools import lru_cache
```
Comment (atalman) [inline torch/xpu/__init__.py:1]: Can you please add some general comments about this package  ? Please refer to cuda, mps __init__.py files

File: test/test_xpu.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+        with torch.xpu.device(target_device):
+            self.assertEqual(target_device, torch.xpu.current_device())
+        self.assertEqual(current_device, torch.xpu.current_device())
+
+        with torch.xpu._DeviceGuard(target_device):
+            self.assertEqual(target_device, torch.xpu.current_device())
+        self.assertEqual(current_device, torch.xpu.current_device())
+
+    def test_get_device_name(self):
```
Comment (atalman) [inline test/test_xpu.py:35]: Looks like following methods are not covered by this test. Is there a possibility to include these tests ?
```
device_of
get_device_capability
get_device_properties
```

File: torch/xpu/__init__.py (hunk 4 above)
Comment (guangyey) [reply torch/xpu/__init__.py:7]: Done.

File: test/test_xpu.py (hunk 5 above)
Comment (guangyey) [reply test/test_xpu.py]: I refine UT to cover `get_device_capability` and `get_device_properties`. Due to the lack of `Allocator` runtime, we can NOT cover `device_of`. But we cover `torch.xpu.device`, a parent class of `device_of`. In my opinion, it is enough.

**Comment by EikanWang** [comment]
@malfet , @atalman , may I know if you have additional comments?

File: .lintrunner.toml [hunk 6]
```diff
@@ -288,6 +288,7 @@ exclude_patterns = [
     'torch/csrc/lazy/**/*',
     'torch/csrc/onnx/init.cpp',
     'torch/csrc/mps/**/*',
+    'torch/csrc/xpu/**',
```
Comment (malfet) [inline .lintrunner.toml:291]: Why do you want to exclude it from lintrunner?

File: aten/src/ATen/test/xpu_device_test.cpp [hunk 7]
```diff
@@ -3,6 +3,7 @@
 #include <aten/src/ATen/xpu/XPUContext.h>
 #include <aten/src/ATen/xpu/XPUDevice.h>
 #include <c10/xpu/XPUFunctions.h>
+#include <torch/torch.h>
```
Comment (malfet) [inline aten/src/ATen/test/xpu_device_test.cpp:6]: Nit (i.e. either include top level meta header, or keep including individual ones)
```suggestion
#include <torch/torch.h>
```

File: torch/csrc/Module.cpp [hunk 8]
```diff
@@ -1462,6 +1469,9 @@ PyObject* initModule() {
 #ifdef USE_CUDA
   THPUtils_addPyMethodDefs(methods, THCPModule_methods());
 #endif
+#ifdef USE_XPU
+  THPUtils_addPyMethodDefs(methods, THXPModule_methods());
+#endif
```
Comment (malfet) [inline torch/csrc/Module.cpp:1474]: Just curious, why we want new TPHXP modules?

File: torch/csrc/api/include/torch/xpu.h [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include <torch/csrc/Export.h>
+
+#include <cstddef>
+#include <cstdint>
+
+namespace torch {
+namespace xpu {
```
Comment (malfet) [inline torch/csrc/api/include/torch/xpu.h:9]: Use nested namespaces please
```suggestion
namespace torch::xpu {
```

File: torch/csrc/api/include/torch/xpu.h [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+/// Returns the number of XPU devices available.
+size_t TORCH_API device_count();
+
+/// Returns true if at least one XPU device is available.
+bool TORCH_API is_available();
+
+} // namespace xpu
+} // namespace torch
```
Comment (malfet) [inline torch/csrc/api/include/torch/xpu.h:18]: Same as above
```suggestion
} // namespace torch::xpu
```

File: torch/csrc/api/src/xpu.cpp [hunk 11]
```diff
@@ -0,0 +1,19 @@
+#include <torch/xpu.h>
+
+#include <ATen/Context.h>
+
+#include <cstddef>
```
Comment (malfet) [inline torch/csrc/api/src/xpu.cpp:5]: Why this is needed (already included from xpu.h isn't it?)

File: torch/csrc/api/src/xpu.cpp [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace torch {
+namespace xpu {
```
Comment (malfet) [inline torch/csrc/api/src/xpu.cpp:8]: Nested namespace plz

File: torch/xpu/__init__.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+from typing import Any, Dict, Optional, Union
+
+import torch
+import torch._C
+from .. import device as _device
+from ._utils import _dummy_type, _get_device_index
+
+_device_t = Union[_device, str, int, None]
+has_half: bool = True
```
Comment (malfet) [inline torch/xpu/__init__.py:16]: Why this is needed?

**Review by malfet — CHANGES_REQUESTED** [review]
I could not choose between "Approve" and "Request Changes" for a really long time.

What is the motivation behind those `torch.xpu.` APIs rather than being identical to `torch.cuda.`? 

For example, what is the motivation behind `torch.xpu.device` or `torch.xpu.device_of` classes when this is already provided by generic `torch.device`?

File: .lintrunner.toml (hunk 6 above)
Comment (guangyey) [reply .lintrunner.toml]: For this PR, there is no clangtidy error. So we remove this line from the exclude list.

File: torch/csrc/Module.cpp (hunk 8 above)
Comment (guangyey) [reply torch/csrc/Module.cpp:1474]: We declare the C++ impl in `THXPModule_methods` and bind them to python APIs here.

File: torch/xpu/__init__.py (hunk 13 above)
Comment (guangyey) [reply torch/xpu/__init__.py:16]: Align to other backends, like `torch.cuda.has_half` is a query API to check if the device supports the half data type. For Intel GPU, it is always `True`.

File: torch/csrc/api/include/torch/xpu.h (hunk 9 above)
Comment (guangyey) [reply torch/csrc/api/include/torch/xpu.h]: Done.

File: torch/csrc/api/src/xpu.cpp (hunk 12 above)
Comment (guangyey) [reply torch/csrc/api/src/xpu.cpp]: Sorry about this.

File: torch/csrc/api/src/xpu.cpp (hunk 11 above)
Comment (guangyey) [reply torch/csrc/api/src/xpu.cpp]: Removed.

File: aten/src/ATen/test/xpu_device_test.cpp (hunk 7 above)
Comment (guangyey) [reply aten/src/ATen/test/xpu_device_test.cpp]: `torch/torch.h` does not include `ATen/xpu/*.h`, so we have to include `ATen/xpu/*`. Here we refine code as below:
```C++
#include <ATen/xpu/XPUContext.h>
#include <ATen/xpu/XPUDevice.h>
#include <torch/torch.h>
```

File: torch/xpu/__init__.py (hunk 13 above)
Comment (malfet) [reply torch/xpu/__init__.py:16]: Correct me if I'm wrong, but there are no XPU devices that do not support half? Then what is the point of adding this property?

File: torch/csrc/Module.cpp (hunk 8 above)
Comment (malfet) [reply torch/csrc/Module.cpp:1474]: TH references old method of implementing kernels, which heavily relied on templates. If I'm not mistaken, we finally got rid of it, but it still exists in one or two places In the codebase

**Comment by EikanWang** [comment]
> I could not choose between "Approve" and "Request Changes" for a really long time.
> 
> What is the motivation behind those `torch.xpu.` APIs rather than being identical to `torch.cuda.`?
> 
> For example, what is the motivation behind `torch.xpu.device` or `torch.xpu.device_of` classes when this is already provided by generic `torch.device`?

Thanks for the comments, @malfet . I'm not sure if I have captured your point. 

The APIs of `torch.xpu` are aligned with `torch.cuda` and almost 1:1 mapping. Regarding the `torch.xpu.device` and `torch.xpu.device_of`, the two classes are provided by `cuda` as well. And the semantics of the two APIs are kind of different from `torch.device`. Take `__enter__` as an example, `with torch.device` stacks the `TorchFunctionMode` while `with torch.cuda.device` only stacks the CUDA device index.

-  `with torch.device("cuda:0")` will create a new `DeviceContext` as a new `TorchFunctionMode` mode to manage the operations requiring device information. It stacks the `TorchFunctionMode` regardless of the `device` of `DeviceContext` being CPU and GPU - https://github.com/pytorch/pytorch/pull/91525, https://github.com/pytorch/pytorch/blob/main/torch/csrc/Device.cpp#L180-L191, https://github.com/pytorch/pytorch/blob/main/torch/utils/_device.py#L59-L78

- `with torch.cuda.device("cuda:0")` is different from the `with torch.device("cuda:0")`. It just impacts the CUDA and stacks the CUDA device index - https://github.com/pytorch/pytorch/blob/main/torch/cuda/__init__.py#L377-L394

Please correct me if my understanding is not accurate.

File: torch/csrc/Module.cpp (hunk 8 above)
Comment (guangyey) [reply torch/csrc/Module.cpp]: Got your point. Use `torch::xpu::python_functions` to replace the old code style.

File: torch/xpu/__init__.py (hunk 13 above)
Comment (guangyey) [reply torch/xpu/__init__.py]: > Correct me if I'm wrong, but there are no XPU devices that do not support half?

Currently, yes it is.
According to your [PR](https://github.com/pytorch/pytorch/pull/115884), PyTorch intends to deprecate this query API in torch 2.3. We remove it here.

**Comment by EikanWang** [comment]
@malfet , may I know if we have addressed comments?

File: torch/xpu/__init__.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+def is_available() -> bool:
+    r"""Return a bool indicating if XPU is currently available."""
+    # This function nerver throws.
+    return device_count() > 0
+
+
+def is_bf16_supported():
+    r"""Return a bool indicating if the current XPU device supports dtype bfloat16."""
+    return True
```
Comment (malfet) [inline torch/xpu/__init__.py:59]: If all XPU devices support bf16, why this flag is needed?

**Review by malfet — APPROVED** [review]
Thank you for all your changes, looks good to me, but I wonder if even fewer methods are really needed here.
Also, please edit PR description to match list of functions you are adding here

**Comment by guangyey** [comment]
> Thank you for all your changes, looks good to me, but I wonder if even fewer methods are really needed here. Also, please edit PR description to match list of functions you are adding here

Thanks for your comments. These methods cannot be minimized further, they are frequently used. We expect to have our API align one-to-one with CUDA, reducing the effort for users to migrate their code.

File: torch/xpu/__init__.py (hunk 14 above)
Comment (guangyey) [reply torch/xpu/__init__.py:59]: Currently, all XPU devices support bf16. We will refine this code when there is a new device breaks this rule. We hope this method can align to 'torch.cuda.is_bf16_supported`, reducing the effort to migrate the user's code.
